core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory2.java

```diff
@@ -238,31 +238,31 @@ private Status getAggregatedError() {
   @VisibleForTesting
   static class Picker extends SubchannelPicker {
     @Nullable
-    final Status status;
+    private final Status status;
     private final List<Subchannel> list;
+    @GuardedBy("this")
     private int index;
-    private final boolean empty;
 
     Picker(List<Subchannel> list, @Nullable Status status) {
-      this.empty = list.isEmpty();
-      this.list = list;
+      this.list = Collections.unmodifiableList(list);
       this.status = status;
     }
 
     @Override
     public PickResult pickSubchannel(Attributes affinity, Metadata headers) {
-      if (!empty) {
+      if (!list.isEmpty()) {
         return PickResult.withSubchannel(nextSubchannel());
-      } else {
-        if (status != null) {
-          return PickResult.withError(status);
-        }
-        return PickResult.withNoResult();
       }
+
+      if (status != null) {
+        return PickResult.withError(status);
+      }
+
+      return PickResult.withNoResult();
     }
 
     private Subchannel nextSubchannel() {
-      if (empty) {
+      if (list.isEmpty()) {
```

Comment:
How about saving `size` as a member variable? This way you can eliminate `isEmpty()` call here and the `list.size()` call on line 271 and 272. In fact line 272 could be `index = 0`.